core/state-machine/src/proving_backend.rs

```diff
@@ -29,6 +30,47 @@ use crate::trie_backend::TrieBackend;
 use crate::trie_backend_essence::{Ephemeral, TrieBackendEssence, TrieBackendStorage};
 use crate::{Error, ExecutionError, Backend};
 
+/// A proof that some set of key-value pairs are included in the storage trie.
+///
+/// The proof consists of the set of serialized nodes in the storage trie accessed when looking up
+/// the keys covered by the proof. Verifying the proof requires constructing the partial trie from
+/// the serialized nodes and performing the key lookups.
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
+pub struct StorageProof {
```

Comment:
I think this should be moved to `substrate-primitives`. State-machine just compiles on native, but the proof stuff is required on `no-std` as well e.g. in Cumulus.

Reply:
OK, I added a `TrieProofWithValues` struct in `substrate-trie` which has a `no_std` option. `StorageProof` in `substrate-state-machine` now wraps that.

The reason these need to be separate is because a `TrieBackend` is an abstration over a single trie which is defined in `substrate-state-machine`. In particular, it allows storage accesses to child tries which traverse multiple tries, a concept foreign to `substrate-trie`.

Currently, these structs are both very simple and are basically just a vec of encoded nodes. In the future, though, it would be nice to create trie proofs that omit redundant internal hashes and storage proofs that internally contain multiple separate trie proofs, one for the root trie and one for each child trie.

Reply:
Could we please get a method `into_inner()` or similar?

Reply:
OK, yes, I removed the whole TrieProof thing and there's a `StorageProof::iter_nodes(self)` method which iterates over all trie nodes.